src/hotspot/os/linux/hugepages.cpp

```diff
@@ -227,15 +228,101 @@ void THPSupport::print_on(outputStream* os) {
+  assert(_initialized, "Not initialized");
+  return _mode;
+}
+
+bool ShmemTHPSupport::is_forced() const {
+  return _mode == ShmemTHPMode::always || _mode == ShmemTHPMode::force || _mode == ShmemTHPMode::within_size;
+}
+
+bool ShmemTHPSupport::is_enabled() const {
+  return is_forced() || _mode == ShmemTHPMode::advise;
+}
+
+bool ShmemTHPSupport::is_disabled() const {
+  return _mode == ShmemTHPMode::never || _mode == ShmemTHPMode::deny || _mode == ShmemTHPMode::unknown;
+}
+
+void ShmemTHPSupport::scan_os() {
+  // Scan /sys/kernel/mm/transparent_hugepage/shmem_enabled
+  // see mm/huge_memory.c
+  _mode = ShmemTHPMode::unknown;
+  const char* filename = "/sys/kernel/mm/transparent_hugepage/shmem_enabled";
+  FILE* f = ::fopen(filename, "r");
+  if (f != nullptr) {
+    char buf[64];
+    char* s = fgets(buf, sizeof(buf), f);
+    assert(s == buf, "Should have worked");
+    if (::strstr(buf, "[always]") != nullptr) {
+      _mode = ShmemTHPMode::always;
+    } else if (::strstr(buf, "[within_size]") != nullptr) {
+      _mode = ShmemTHPMode::within_size;
+    } else if (::strstr(buf, "[advise]") != nullptr) {
+      _mode = ShmemTHPMode::advise;
+    } else if (::strstr(buf, "[never]") != nullptr) {
+      _mode = ShmemTHPMode::never;
+    } else if (::strstr(buf, "[deny]") != nullptr) {
+      _mode = ShmemTHPMode::deny;
+    } else if (::strstr(buf, "[force]") != nullptr) {
+      _mode = ShmemTHPMode::force;
+    } else {
+      assert(false, "Weird content of %s: %s", filename, buf);
+    }
+    fclose(f);
+  }
+
+  _initialized = true;
+
+  LogTarget(Info, pagesize) lt;
+  if (lt.is_enabled()) {
+    LogStream ls(lt);
+    print_on(&ls);
+  }
+}
+
+const char* ShmemTHPSupport::mode_to_string(ShmemTHPMode mode) {
+  switch (mode) {
+    case ShmemTHPMode::always:      return "always";
+    case ShmemTHPMode::advise:      return "advise";
+    case ShmemTHPMode::within_size: return "within_size";
+    case ShmemTHPMode::never:       return "never";
+    case ShmemTHPMode::deny:        return "deny";
+    case ShmemTHPMode::force:       return "force";
+    case ShmemTHPMode::unknown:      // Fallthrough
+    default:                        return "unknown";
+  };
+}
+
+void ShmemTHPSupport::print_on(outputStream* os) {
+  if (_initialized) {
+    os->print_cr("Shared memory transparent hugepage (THP) support:");
+    os->print_cr("  Shared memory THP mode: %s", mode_to_string(_mode));
+  } else {
+    os->print_cr("  unknown.");
+  }
+}
+
 StaticHugePageSupport HugePages::_static_hugepage_support;
 THPSupport HugePages::_thp_support;
+ShmemTHPSupport HugePages::_shmem_thp_support;
+bool HugePages::_thp_requested{false};
 
 void HugePages::initialize() {
   _static_hugepage_support.scan_os();
   _thp_support.scan_os();
+  _shmem_thp_support.scan_os();
+
+  const bool huge_pages_turned_off = !FLAG_IS_DEFAULT(UseLargePages) && !UseLargePages;
+  _thp_requested = UseTransparentHugePages && !huge_pages_turned_off;
```

Comment:
This muddles the water a bit, since the original intent of HugePages vs whatever happens in os_linux was to let HugePages give me the unadulterated info of what the OS supports, whereas processing switches and deciding on them should happen in os_linux in large_page_init. Would it be possible to move "_thp_requested" up to the caller?

We can keep the "should_madvise_anonymous_thps" since those make sense here, but move the "requested" condition up to the caller.